finalfrontier/src/idx.rs

```diff
@@ -0,0 +1,114 @@
+use std::option;
+
+/// A single index in the
+#[derive(Clone)]
+pub struct SingleIdx {
+    word_idx: u64,
+}
+
+#[derive(Clone)]
+pub struct WordWithSubwordsIdx {
+    word_idx: u64,
+    subwords: Vec<u64>,
+}
+
+impl WordWithSubwordsIdx {
+    pub(crate) fn new(word_idx: u64, subwords: impl Into<Vec<u64>>) -> Self {
+        WordWithSubwordsIdx {
+            word_idx,
+            subwords: subwords.into(),
+        }
+    }
+}
+
+/// Index representation.
+///
+/// This enum represents possible types of indices, e.g. in a vocabulary with subwords, the `Multi`
+/// variant includes all indices associated with a key.
+pub trait WordIdx: Clone {
+    /// Return the unique word index for the WordIdx.
+    fn word_idx(&self) -> u64;
+
+    /// Build a new WordIdx containing only a single index.
+    fn from_word_idx(word_idx: u64) -> Self;
+}
+
+#[allow(clippy::len_without_is_empty)]
+pub trait IndexIterator {
+    type Iter: Iterator<Item = u64>;
+
+    /// Return the number of indices.
+    fn len(&self) -> usize;
+
+    /// Return an iterator over the indices.
+    fn iter(&self) -> Self::Iter;
```

Comment:
Anything against implementing `IntoIterator` instead? `len` could then perhaps move to the `WordIdx` trait?

Reply:
That is actually more complicated since we're only passing references to the embeddings method. If we move the Len to the other trait then we'd need to implement it on references, too. 

There's a workaround by counting the indices in the for loop but that's not nice either.

Reply:
I don't follow. Is the problem `IntoIterator` (which could be implemented  for e.g. `&SingleIdx`) or the `len` method? Why couldn't `len` be a method of `WordIdx` without implementations on references? See e.g.:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=803038d912d7e06ed4fb565b7431ab26

Reply:
Turns out I was wrong. I'm implementing it with `IntoIterator` now and I'm ditching the `IndexIterator` trait.